### PR TITLE
Enable exodus restart for mesh generator system

### LINKED ISSUE #15554: Mesh generator does not have the Exodus file restart capability
## Reason
With the mesh generator system, particularly with `FileMeshGenerator`, we no longer have the support for restarting variables from an Exodus file. Although I personally do not like this capability, it is needed for applications to move from mesh modifiers to mesh generators. 

## Design
We have to let the `MooseApp` to own the `ExodusII_IO`, which is currently in `MooseMesh`. This makes sense because it is `MooseApp` requesting the restart from Exodus file. So we can let `FileMesh` to hand over the exodus reader to `MooseApp`. This change will enable us to do the same in `FileMeshGenerator`. Because we cannot assume there will be only one mesh generator, we possibly should have a control parameter `use_for_exodus_restart` to control if a `FileMeshGenerator` is suitable for variable restarting.

This is needed by Griffin for using the mesh generators. So I will work on this and send in a PR.

## Impact
Keep capability in the new design.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #15554.

### CODE REVIEW

**Comment by YaqiWang** [comment]
This change will break Griffin. Not sure if it will break modules or other apps.

**Comment by YaqiWang** [comment]
Looks like that else check in `FEProblemBase::initialSetup` did catch something. I am looking into the Pronghorn failing test.

**Comment by YaqiWang** [comment]
Pronghorn can be patched without this PR. I've sent in a MR in there. I might roll back the old interface in `MoosApp` to avoid breaks in Moltres. It will be your call. This is ready for review.

File: framework/src/meshgenerators/FileMeshGenerator.C [hunk 1]
```diff
@@ -28,6 +28,10 @@ FileMeshGenerator::validParams()
   params.addParam<std::vector<std::string>>(
       "exodus_extra_element_integers",
       "The variable names in the mesh file for loading extra element integers");
+  params.addParam<bool>("use_for_exodus_restart",
+                        false,
+                        "True to indicate that the mesh file this generator is used can be used "
+                        "for restarting variables");
```
Comment (lindsayad) [inline framework/src/meshgenerators/FileMeshGenerator.C:34]: This doc string is confusing. Maybe "True to indicate that the mesh file this generator is reading can be used for restarting variables"?

**Review by lindsayad — CHANGES_REQUESTED** [review]
A lot of the name changes in this PR go from a general file or reader name to a specific exodus name. Would it be more far sighted to keep the general names or are we confident that the types of restart being done here will always be exodus specific?

**Comment by YaqiWang** [comment]
I think the naming now is better because this capability is indeed only for exodus. If we have supports with other format in the future, we can change later.

File: framework/src/meshgenerators/FileMeshGenerator.C (hunk 1 above)
Comment (YaqiWang) [reply framework/src/meshgenerators/FileMeshGenerator.C:34]: Sure.

**Comment by lindsayad** [comment]
> I think the naming now is better because this capability is indeed only for exodus. If we have supports with other format in the future, we can change later.

As long as you're future volunteering to fix the apps when you change an API, then I'm good with it :smile:

**Review by lindsayad — APPROVED** [review]

**Comment by YaqiWang** [comment]
I will. I guess I have to wait for it at least in `devel` to send in patches, right?

**Comment by lindsayad** [comment]
No you can update the MOOSE submodule to point to next in any broken apps.

**Comment by YaqiWang** [comment]
Great. Btw, I did not see Griffin was tested in app tests. Do you know which target it is in?

**Comment by YaqiWang** [comment]
I have the patch for Moltres ready. Just waiting for moose submodule update. I will patch Griffin as well. Need to be away for about 2 or 3 hours. So feel free to merge this a little later.

**Pull request merged**